File: scripts/processor/events.py
"""
events.py – Event loading, debouncing, coordinate mapping, and interpolation.

All functions are pure (no side effects, no global state) and operate on
plain dicts / lists, keeping them easy to test and reason about.
"""

from __future__ import annotations

import bisect
import json
import os
import sys
from typing import Any


# ─── Type aliases (plain dicts — we don't need heavy dataclasses here) ───────

Event = dict[str, Any]


# ─── Loading ──────────────────────────────────────────────────────────────────

def load_events(path: str) -> list[Event]:
    """Load events.json.  Returns [] on missing or invalid file."""
    if not path or not os.path.exists(path):
        return []
    try:
        with open(path, "r") as f:
            data = json.load(f)
        return data if isinstance(data, list) else []
    except Exception as exc:
        print(f"[Events] Warning: failed to load {path}: {exc}", file=sys.stderr)
        return []


def split_events(
    events: list[Event],
) -> tuple[list[Event], list[Event], list[Event]]:
    """Split raw events into (clicks, scrolls, moves), each sorted by timestamp."""
    clicks = sorted(
        [e for e in events if e.get("type") == "click"],
        key=lambda e: e["timestamp"],
    )
    scrolls = sorted(
        [e for e in events if e.get("type") == "scroll"],
        key=lambda e: e["timestamp"],
    )
    moves = sorted(
        [e for e in events if e.get("type") == "move"],
        key=lambda e: e["timestamp"],
    )
    return clicks, scrolls, moves


# ─── Click debouncing ────────────────────────────────────────────────────────

def debounce_clicks(clicks: list[Event], gap: float = 0.4) -> list[Event]:
    """Merge rapid-fire click bursts into a single event.

    Within each burst (gap ≤ *gap* seconds) the **last** click is kept so
    that the zoom target lands on the final position of the burst.
    """
    if not clicks:
        return []

    result: list[Event] = []
    burst: list[Event] = [clicks[0]]

    for click in clicks[1:]:
        if click["timestamp"] - burst[-1]["timestamp"] <= gap:
            burst.append(click)
        else:
            result.append(burst[-1])
            burst = [click]
    result.append(burst[-1])
    return result


# ─── Coordinate mapping ──────────────────────────────────────────────────────

def to_canvas_coords(
    events: list[Event],
    origin_x: float,
    origin_y: float,
    scale: float,
    padding: int,
    frame_w: int,
    frame_h: int,
) -> list[Event]:
    """Translate global screen coords → canvas-space coords.

    Pipeline:  (event - origin) * scale + padding, clamped to recording area.
    """
    out: list[Event] = []
    for e in events:
        ne = dict(e)
        if "x" in ne:
            ne["x"] = (ne["x"] - origin_x) * scale + padding
            ne["x"] = max(padding, min(frame_w + padding, ne["x"]))
        if "y" in ne:
            ne["y"] = (ne["y"] - origin_y) * scale + padding
            ne["y"] = max(padding, min(frame_h + padding, ne["y"]))
        out.append(ne)
    return out


# ─── Meta loading ────────────────────────────────────────────────────────────

def load_meta(
    meta_path: str | None, frame_w: int
) -> tuple[float, float, float]:
    """Parse meta.json → (origin_x, origin_y, scale_factor).

    Returns (0, 0, 1.0) when meta is unavailable or broken.
    """
    if not meta_path or not os.path.exists(meta_path):
        return 0.0, 0.0, 1.0

    try:
        with open(meta_path, "r") as f:
            meta = json.load(f)

        ox = float(meta.get("originX", 0))
        oy = float(meta.get("originY", 0))

        sf = meta.get("scaleFactor")
        cw = meta.get("captureWidth")

        if sf and float(sf) > 0:
            scale = float(sf)
        elif cw and int(cw) > 0:
            scale = frame_w / int(cw)
        else:
            scale = 1.0

        source = meta.get("sourceType", "unknown")
        print(
            f"[Events] Capture origin=({ox}, {oy}), "
            f"source={source}, scale={scale:.4f}",
            file=sys.stderr,
        )
        return ox, oy, scale
    except Exception as exc:
        print(f"[Events] Warning: meta.json parse failed: {exc}", file=sys.stderr)
        return 0.0, 0.0, 1.0


# ─── Cursor interpolation ────────────────────────────────────────────────────

class CursorInterpolator:
    """Fast cursor-position lookup with pre-built timestamp index.

    Uses bisect for O(log n) timestamp lookup instead of linear scan.
    """

    def __init__(self, moves: list[Event]) -> None:
        self._moves = moves
        self._times = [m["timestamp"] for m in moves]

    def at(self, t: float) -> tuple[float, float] | None:
        """Return interpolated (x, y) at time *t*, or None if no data."""
        if not self._moves:
            return None

        if t <= self._times[0]:
            return self._moves[0]["x"], self._moves[0]["y"]

        if t >= self._times[-1]:
            return self._moves[-1]["x"], self._moves[-1]["y"]

        # Find insertion point → bracketing pair is (idx-1, idx)
        idx = bisect.bisect_right(self._times, t)
        if idx <= 0:
            idx = 1
        if idx >= len(self._moves):
            idx = len(self._moves) - 1

        a = self._moves[idx - 1]
        b = self._moves[idx]
        dt = b["timestamp"] - a["timestamp"]
        if dt <= 0:
            return a["x"], a["y"]

        frac = (t - a["timestamp"]) / dt
        x = a["x"] + (b["x"] - a["x"]) * frac
        y = a["y"] + (b["y"] - a["y"]) * frac
        return x, y


# ─── Camera scheduling ───────────────────────────────────────────────────────

def schedule_camera(
    camera,  # SmoothCamera (avoid circular import)
    clicks: list[Event],
    scrolls: list[Event],
    cursor: CursorInterpolator,
    t: float,
    zoom_factor: float,
    hold_duration: float,
) -> None:
    """Set camera target + zoom for the current time.

    Rules:
      1. If a click is active (within hold window): zoom in, snap to click,
         then gently follow cursor after a short settle period.
      2. Otherwise: zoom out, lazily recenter.
      3. Scroll events add a temporary vertical offset.
    """
    active_click: Event | None = None
    click_age = 0.0

    for click in clicks:
        ct = click["timestamp"]
        if ct <= t <= ct + hold_duration:
            active_click = click
            click_age = t - ct

    if active_click is not None:
        camera.set_zoom(zoom_factor)

        settle = 0.35  # seconds before switching to cursor-follow
        if click_age < settle:
            camera.set_target(active_click["x"], active_click["y"], mode="snap")
        else:
            pos = cursor.at(t)
            if pos:
                camera.set_target(pos[0], pos[1], mode="follow")
            else:
                camera.set_target(active_click["x"], active_click["y"], mode="snap")
    else:
        camera.set_target(camera.canvas_w / 2.0, camera.canvas_h / 2.0, mode="recenter")
        camera.reset_zoom()

    # Scroll offset (temporary vertical nudge)
    for scroll in scrolls:
        st = scroll["timestamp"]
        dur = 0.6
        if st <= t <= st + dur:
            progress = (t - st) / dur
            ease = progress * progress * (3.0 - 2.0 * progress)  # smoothstep
            offset = scroll.get("rotation", 0) * 55.0 * (1.0 - ease)
            camera._target_y = max(
                0.0, min(float(camera.canvas_h), camera._target_y + offset)
            )

File: src/renderer/lib/zoom-engine/events.ts
// events.ts — Event processing, coordinate mapping, cursor interpolation,
// and Screen-Studio-style camera scheduling.

import type { InputEvent, ClickEvent, ScrollEvent, MoveEvent } from '../../../shared/types';
import type { ZoomSegment } from './segments';

// ─── Splitting ─────────────────────────────────────────────────────────────

export function splitEvents(events: InputEvent[]): {
  clicks: ClickEvent[];
  scrolls: ScrollEvent[];
  moves: MoveEvent[];
} {
  const clicks: ClickEvent[] = [];
  const scrolls: ScrollEvent[] = [];
  const moves: MoveEvent[] = [];

  for (const e of events) {
    if (e.type === 'click') clicks.push(e);
    else if (e.type === 'scroll') scrolls.push(e);
    else if (e.type === 'move') moves.push(e);
  }

  clicks.sort((a, b) => a.timestamp - b.timestamp);
  scrolls.sort((a, b) => a.timestamp - b.timestamp);
  moves.sort((a, b) => a.timestamp - b.timestamp);

  return { clicks, scrolls, moves };
}

// ─── Click debouncing ──────────────────────────────────────────────────────

export function debounceClicks(clicks: ClickEvent[], gap = 0.4): ClickEvent[] {
  if (clicks.length === 0) return [];

  const result: ClickEvent[] = [];
  let burst: ClickEvent[] = [clicks[0]];

  for (let i = 1; i < clicks.length; i++) {
    if (clicks[i].timestamp - burst[burst.length - 1].timestamp <= gap) {
      burst.push(clicks[i]);
    } else {
      result.push(burst[burst.length - 1]);
      burst = [clicks[i]];
    }
  }
  result.push(burst[burst.length - 1]);
  return result;
}

// ─── Coordinate mapping ───────────────────────────────────────────────────

export function toCanvasCoords<T extends InputEvent>(
  events: T[],
  originX: number,
  originY: number,
  scale: number,
  padding: number,
  frameW: number,
  frameH: number,
): T[] {
  return events.map((e) => {
    const ne = { ...e };
    if ('x' in ne && typeof ne.x === 'number') {
      ne.x = (ne.x - originX) * scale + padding;
      ne.x = Math.max(padding, Math.min(frameW + padding, ne.x));
    }
    if ('y' in ne && typeof ne.y === 'number') {
      ne.y = (ne.y - originY) * scale + padding;
      ne.y = Math.max(padding, Math.min(frameH + padding, ne.y));
    }
    return ne;
  });
}

// ─── Meta loading ──────────────────────────────────────────────────────────

export interface MetaInfo {
  originX: number;
  originY: number;
  scaleFactor: number;
}

export function parseMeta(
  meta: { originX?: number; originY?: number; scaleFactor?: number; captureWidth?: number } | null,
  frameW: number,
): MetaInfo {
  if (!meta) return { originX: 0, originY: 0, scaleFactor: 1 };

  const originX = meta.originX ?? 0;
  const originY = meta.originY ?? 0;

  let scaleFactor = 1;
  if (meta.scaleFactor && meta.scaleFactor > 0) {
    scaleFactor = meta.scaleFactor;
  } else if (meta.captureWidth && meta.captureWidth > 0) {
    scaleFactor = frameW / meta.captureWidth;
  }

  return { originX, originY, scaleFactor };
}

// ─── Cursor interpolation ──────────────────────────────────────────────────

export class CursorInterpolator {
  private _moves: MoveEvent[];
  private _times: number[];

  constructor(moves: MoveEvent[]) {
    this._moves = moves;
    this._times = moves.map((m) => m.timestamp);
  }

  at(t: number): { x: number; y: number } | null {
    if (this._moves.length === 0) return null;

    if (t <= this._times[0]) {
      return { x: this._moves[0].x, y: this._moves[0].y };
    }

    if (t >= this._times[this._times.length - 1]) {
      const last = this._moves[this._moves.length - 1];
      return { x: last.x, y: last.y };
    }

    // Binary search for insertion point
    let lo = 0;
    let hi = this._times.length;
    while (lo < hi) {
      const mid = (lo + hi) >>> 1;
      if (this._times[mid] <= t) lo = mid + 1;
      else hi = mid;
    }

    let idx = lo;
    if (idx <= 0) idx = 1;
    if (idx >= this._moves.length) idx = this._moves.length - 1;

    const a = this._moves[idx - 1];
    const b = this._moves[idx];
    const dt = b.timestamp - a.timestamp;
    if (dt <= 0) return { x: a.x, y: a.y };

    const frac = (t - a.timestamp) / dt;
    return {
      x: a.x + (b.x - a.x) * frac,
      y: a.y + (b.y - a.y) * frac,
    };
  }
}

// ─── Camera scheduling (Screen Studio style) ──────────────────────────────
//
// Uses pre-computed zoom segments for smart zoom decisions:
//   1. Zoom in when entering a segment → snap camera to click position
//   2. While zoomed, follow cursor with lazy lerp + dead zone
//   3. Zoom out smoothly when leaving a segment
//   4. Anticipate upcoming segments — don't zoom out if one is close
//   5. Scroll nudges camera vertically while zoomed
//
// The key: fewer, longer zoom holds with slow transitions = Screen Studio feel.

const DEAD_ZONE = 45;      // px — ignore cursor movements smaller than this (relative to actual cam position)
const SNAP_PHASE = 0.4;    // seconds — snap to click position on new click
const ANTICIPATION = 0.8;  // seconds — don't zoom out if next segment is this close

export function scheduleCamera(
  camera: import('./spring').SmoothCamera,
  segments: ZoomSegment[],
  clicks: ClickEvent[],
  scrolls: ScrollEvent[],
  cursor: CursorInterpolator,
  t: number,
  zoomFactor: number,
): void {
  // ── Find active segment and next upcoming segment ────────────
  let activeSegment: ZoomSegment | null = null;
  let nextSegment: ZoomSegment | null = null;

  for (const seg of segments) {
    if (t >= seg.startTime && t <= seg.endTime) {
      activeSegment = seg;
    } else if (seg.startTime > t && !nextSegment) {
      nextSegment = seg;
    }
  }

  // ── Find most recent click within current segment ────────────
  let latestClick: ClickEvent | null = null;
  if (activeSegment) {
    for (const click of clicks) {
      if (click.timestamp <= t && click.timestamp >= activeSegment.startTime) {
        latestClick = click;
      }
    }
  }

  if (activeSegment) {
    // ── Zoomed in ─────────────────────────────────────────────
    camera.setZoom(zoomFactor);

    const timeSinceClick = latestClick ? t - latestClick.timestamp : Infinity;

    if (latestClick && timeSinceClick < SNAP_PHASE) {
      // Just clicked — move toward click position
      camera.setTarget(latestClick.x, latestClick.y, 'snap');
    } else {
      // Follow cursor with lazy lerp + dead zone against ACTUAL camera position
      const pos = cursor.at(t);
      if (pos) {
        // Compare to actual camera position, not target — prevents dead zone oscillation
        const dx = pos.x - camera.x;
        const dy = pos.y - camera.y;
        const dist = Math.sqrt(dx * dx + dy * dy);

        if (dist > DEAD_ZONE) {
          camera.setTarget(pos.x, pos.y, 'follow');
        }
      } else if (latestClick) {
        camera.setTarget(latestClick.x, latestClick.y, 'follow');
      }
    }

    // ── Scroll offset (computed once, not additive) ────────────
    // Find the strongest active scroll and apply a single offset
    let scrollOffset = 0;
    for (const scroll of scrolls) {
      const dur = 0.8;
      if (scroll.timestamp <= t && t <= scroll.timestamp + dur) {
        const progress = (t - scroll.timestamp) / dur;
        const ease = progress * progress * (3 - 2 * progress);
        const offset = (scroll.rotation ?? 0) * 12 * (1 - ease);
        // Keep the largest magnitude offset (not cumulative)
        if (Math.abs(offset) > Math.abs(scrollOffset)) scrollOffset = offset;
      }
    }
    if (scrollOffset !== 0) {
      camera.targetY = Math.max(0, Math.min(camera.canvasH, camera.targetY + scrollOffset));
    }
  } else if (nextSegment && (nextSegment.startTime - t) < ANTICIPATION) {
    // ── About to enter a segment — hold position, don't zoom out
  } else {
    // ── Not zoomed — drift back to center ─────────────────────
    camera.resetZoom();
    camera.setTarget(camera.canvasW / 2, camera.canvasH / 2, 'recenter');
  }
}
